File: 左神2017/1.排序/8.基数排序/RadixSort.java
public class RadixSort{
    public static void radixSort(int[] arr, int begin, int end, int digit){
        final int radix = 10;
        int i = 0, j = 0,
        int[] count = new int[radix];
        int[] bucket = new int[end-begin+1];
        for (int d = 1; d <= digit; d++){
            for(i=0; i<radix; i++){
                j = getDigit(arr[i], d);
                count[j]++;
            }

            for(i=begin;i<=end;i++){
                count[i] = count[i] + count[i-1];
            }

            for(i=end; i >=begin; i--){
                j = getDigit(arr[i], d);
                bucket[count[j] - 1] = arr[i];
                count[j]--;
            }

            for(i=begin,j=0; i<=end, i++,j++){
                arr[i] = bucket[j];
            }
        }
    }

    public static int getDigit(int x, int d){
        return ((x/((int) Math.pow(10, d-1)) % 10));
    }

}

File: 左神2017/1.排序/8.基数排序/RadixSort.py
def radix_sort(arr, radix=10):
    """a为整数列表， radix为基数"""
    j = len(str(max(arr)))
    i = 0
    while i < j:
        # 初始化桶数组，不能使用 [[]]*radix，这样会共享[]的地址，造成数据重复
        bucket = [[] for _ in range(radix)]
        for x in arr:
            bucket[int(x / (10 ** i) % 10)].append(x)

        # 清空数组，重新压入数据
        arr.clear()
        for x in bucket:
            for y in x:
                arr.append(y)
        i += 1
    return arr


if __name__ == '__main__':
    assert radix_sort([78, 87, 98, 100]) == [78, 87, 98, 100]
